src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs

```diff
@@ -1636,6 +1637,7 @@ public async Task ServerSendsTrailingHeaders_Success()
                 await requestStream.ReadRequestDataAsync();
                 await requestStream.SendResponseAsync(isFinal: false);
                 await requestStream.SendResponseHeadersAsync(null, new[] { new HttpHeaderData("MyHeader", "MyValue") });
+                await semaphore.WaitAsync();
```

Comment:
I'm wondering if we can get stuck here forever if the client side fails for whatever reason. It may be find for the test. But that makes me wonder if we can simply move declaration of the `connection` above the task and perhaps keeping it alive by checking some properties after the client request is finished.

Reply:
No, we won't stuck because, at the end of the test, we're awaiting them with timeout. We can also try something else to do it, but this pattern is commonly used in this test suite.

Reply:
That would make the test fail but the server `Task` would still be blocked, right?

Reply:
Hmm, I think I can also add a timeout to this `WaitAsync` as well, in that case.

Reply:
> That would make the test fail but the server Task would still be blocked, right?

If the await gets never completed, wouldn't the state machine get GC collected anyway, since nothing is rooting it?

Reply:
A `WaitAsync` Task with a timeout will root itself.
The Timer it creates internally has a reference to the Task, while also being in a static field (the list of all timers).

Reply:
But a `WaitAsync` on `SemaphoreSlim` without timeout won't root the state machine, right?

Reply:
The `WaitAsync` by itself won't, but if you're waiting on a `SemaphoreSlim` without a timeout, the expectation is that something else is going to signal you to continue eventually (or you're stuck regardless).

The `Task` returned by `WaitAsync` is stored in a linked list of waiters inside the `SemaphoreSlim`.
So as long as there is something out there with a reference to the semaphore and rooting it, it is by extension rooting the state machine of the `WaitAsync` caller.